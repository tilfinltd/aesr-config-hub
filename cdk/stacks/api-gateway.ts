import { Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
import {
  DomainName,
  RestApi,
  CognitoUserPoolsAuthorizer,
  MockIntegration,
  PassthroughBehavior,
  Model,
  LambdaIntegration,
} from "aws-cdk-lib/aws-apigateway";
import { ICertificate } from "aws-cdk-lib/aws-certificatemanager";
import { IUserPool } from "aws-cdk-lib/aws-cognito";
import * as lambda from "aws-cdk-lib/aws-lambda";
import { ARecord, HostedZone, RecordTarget } from "aws-cdk-lib/aws-route53";
import { ApiGatewayDomain } from "aws-cdk-lib/aws-route53-targets";

export interface ApiGatewayStackProps extends StackProps {
  prefix: string;
  subDomain: string;
  certificate: ICertificate;
  userPool: IUserPool;
  configProviderLambdaName: string;
}

export class ApiGatewayStack extends Stack {
  constructor(scope: Construct, id: string, props: ApiGatewayStackProps) {
    super(scope, id, props);
    const { prefix, subDomain, certificate } = props;

    const rootDomain = `${subDomain}.aesr.dev`;

    const hostedZone = HostedZone.fromLookup(this, "HostedZone", {
      domainName: rootDomain,
    });

    const apiDomain = `api.${subDomain}.aesr.dev`;
    const restApi = new RestApi(this, `${prefix}-api`);

    const domainName = new DomainName(this, "ApiDomain", {
      domainName: apiDomain,
      certificate,
    });

    domainName.addBasePathMapping(restApi);

    new ARecord(this, "ApiDomainRecord", {
      zone: hostedZone,
      recordName: apiDomain,
      target: RecordTarget.fromAlias(new ApiGatewayDomain(domainName)),
    });

    const authorizer = new CognitoUserPoolsAuthorizer(
      this,
      "CognitoAuthorizer",
      {
        authorizerName: "CognitoAuthorizer",
        cognitoUserPools: [props.userPool],
      },
    );

    const callbackRes = restApi.root.addResource("callback");
    callbackRes.addMethod(
      "GET",
      new MockIntegration({
        passthroughBehavior: PassthroughBehavior.WHEN_NO_MATCH,
        requestTemplates: {
          "application/json": '{"statusCode":200}',
        },
        integrationResponses: [
          {
            statusCode: "200",
            responseTemplates: {
              "text/html": `<html>
            <body style="font-family:sans-serif;padding:36px">
             <p style="font-size:18px;line-height:2">
               Please keep this tab open and<br>
               <u>click on the <strong>AWS Extend Switch Roles</strong> icon</u>
               in your browser toolbar<br>
               to complete the connection to <strong>AESR Config Hub</strong>.
             </p>
            </body>
           </html>`,
            },
          },
        ],
      }),
      {
        methodResponses: [
          {
            statusCode: "200",
            responseModels: { "text/html": Model.EMPTY_MODEL },
          },
        ],
      },
    );

    const configProviderLambda = lambda.Function.fromFunctionName(
      this,
      "ConfigProviderLambda",
      props.configProviderLambdaName,
    );
    const userRes = restApi.root.addResource("user");
    const configRes = userRes.addResource("config");
    configRes.addMethod("GET", new LambdaIntegration(configProviderLambda), {
      authorizer,
    });
  }
}
