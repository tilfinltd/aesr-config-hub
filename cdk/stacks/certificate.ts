import { Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
import {
  Certificate,
  CertificateValidation,
  ICertificate,
} from "aws-cdk-lib/aws-certificatemanager";
import { HostedZone } from "aws-cdk-lib/aws-route53";

export interface CertificateStackProps extends StackProps {
  subDomain: string;
}

export class GlobalCertificateStack extends Stack {
  public readonly authCertificate: ICertificate;

  constructor(scope: Construct, id: string, props: CertificateStackProps) {
    super(scope, id, props);

    const rootDomain = `${props.subDomain}.aesr.dev`;

    const hostedZone = HostedZone.fromLookup(this, "HostedZone", {
      domainName: rootDomain,
    });

    this.authCertificate = new Certificate(this, "AuthDomainCertificate", {
      domainName: `auth.${rootDomain}`,
      validation: CertificateValidation.fromDns(hostedZone),
    });
  }
}

export class LocalCertificateStack extends Stack {
  public readonly apiCertificate: ICertificate;

  constructor(scope: Construct, id: string, props: CertificateStackProps) {
    super(scope, id, props);

    const rootDomain = `${props.subDomain}.aesr.dev`;

    const hostedZone = HostedZone.fromLookup(this, "HostedZone", {
      domainName: rootDomain,
    });

    this.apiCertificate = new Certificate(this, "ApiDomainCertificate", {
      domainName: `api.${rootDomain}`,
      validation: CertificateValidation.fromDns(hostedZone),
    });
  }
}
