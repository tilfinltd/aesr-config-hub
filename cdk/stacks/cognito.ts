import { Duration, RemovalPolicy, Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
import { ICertificate } from "aws-cdk-lib/aws-certificatemanager";
import { IUserPool, OAuthScope, UserPool } from "aws-cdk-lib/aws-cognito";
import { ARecord, HostedZone, RecordTarget } from "aws-cdk-lib/aws-route53";
import { UserPoolDomainTarget } from "aws-cdk-lib/aws-route53-targets";

export interface CognitoStackProps extends StackProps {
  prefix: string;
  subDomain: string;
  certificate: ICertificate;
  refreshTokenValidity: Duration;
}

export class CognitoStack extends Stack {
  public readonly userPool: IUserPool;

  constructor(scope: Construct, id: string, props: CognitoStackProps) {
    super(scope, id, props);
    const { prefix, subDomain, certificate } = props;

    const rootDomain = `${subDomain}.aesr.dev`;

    const hostedZone = HostedZone.fromLookup(this, "HostedZone", {
      domainName: rootDomain,
    });

    // For Cognito custom domain, an A record for the root domain must exist.
    new ARecord(this, "RootDomainRecord", {
      zone: hostedZone,
      recordName: rootDomain,
      target: RecordTarget.fromIpAddresses("127.0.0.1"),
    });

    const authDomain = `auth.${subDomain}.aesr.dev`;
    const callbackUrl = `https://api.${subDomain}.aesr.dev/callback`;

    const userPool = new UserPool(this, "UserPool", {
      userPoolName: `${prefix}-user-pool`,
      standardAttributes: {
        email: { required: true, mutable: true },
      },
      removalPolicy: RemovalPolicy.DESTROY,
    });

    const customDomain = userPool.addDomain("UserPoolDomain", {
      customDomain: {
        domainName: authDomain,
        certificate,
      },
    });

    new ARecord(this, "AuthDomainRecord", {
      zone: hostedZone,
      recordName: authDomain,
      target: RecordTarget.fromAlias(new UserPoolDomainTarget(customDomain)),
    });

    userPool.addClient("UserPoolClient", {
      userPoolClientName: `${prefix}-browser-extension`,
      generateSecret: false,
      oAuth: {
        flows: {
          authorizationCodeGrant: true,
        },
        callbackUrls: [callbackUrl],
        scopes: [OAuthScope.OPENID, OAuthScope.EMAIL],
      },
      idTokenValidity: Duration.hours(1),
      refreshTokenValidity: props.refreshTokenValidity,
    });

    this.userPool = userPool;
  }
}
