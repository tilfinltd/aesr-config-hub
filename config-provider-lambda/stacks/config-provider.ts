import { Duration, Stack, StackProps, aws_lambda_nodejs } from "aws-cdk-lib";
import { Construct } from "constructs";
import {
  Effect,
  ManagedPolicy,
  PolicyStatement,
  Role,
  ServicePrincipal,
} from "aws-cdk-lib/aws-iam";
import * as lambda from "aws-cdk-lib/aws-lambda";
import { LogGroup } from "aws-cdk-lib/aws-logs";
import { Bucket } from "aws-cdk-lib/aws-s3";

export interface ConfigProviderStackProps extends StackProps {
  userConfigBucketName: string;
  lambdaName: string;
}

export class ConfigProviderStack extends Stack {
  constructor(scope: Construct, id: string, props: ConfigProviderStackProps) {
    super(scope, id, props);
    const { userConfigBucketName, lambdaName } = props;

    const lambdaRole = new Role(this, "ConfigProviderLambdaExecutionRole", {
      roleName: `${lambdaName}-LambdaExecutionRole`,
      assumedBy: new ServicePrincipal("lambda.amazonaws.com"),
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName(
          "service-role/AWSLambdaBasicExecutionRole",
        ),
      ],
    });

    new aws_lambda_nodejs.NodejsFunction(this, "ConfigProviderLambda", {
      functionName: lambdaName,
      runtime: lambda.Runtime.NODEJS_20_X,
      bundling: {
        externalModules: [
          '@aws-sdk/*', // Use the AWS SDK for JS v3 available in the Lambda runtime
        ],
      },
      entry: "lambda/index.ts",
      handler: "handler",
      memorySize: 128,
      timeout: Duration.seconds(10),
      role: lambdaRole,
      environment: {
        USER_CONFIG_BUCKET: userConfigBucketName,
      },
    });

    const logGroup = new LogGroup(this, "ConfigProviderLogGroup", {
      logGroupName: `/aws/lambda/${lambdaName}`,
    });
    logGroup.grantWrite(lambdaRole);

    const userConfigBucket = new Bucket(this, "UserConfigBucket", {
      bucketName: userConfigBucketName,
    });

    userConfigBucket.addToResourcePolicy(
      new PolicyStatement({
        effect: Effect.ALLOW,
        actions: ["s3:GetObject"],
        principals: [lambdaRole],
        resources: [userConfigBucket.bucketArn + "/*"],
      }),
    );
  }
}
