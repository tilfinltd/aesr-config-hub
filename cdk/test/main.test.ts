import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import {
  GlobalCertificateStack,
  LocalCertificateStack,
} from "../stacks/certificate";
import { CognitoStack } from "../stacks/cognito";
import { ApiGatewayStack } from "../stacks/api-gateway";

const cdkJson = require("../cdk.json");

describe("All stacks", () => {
  test("default", () => {
    const app = new App(cdkJson);
    const env = {
      region: "ap-northeast-1",
      account: "000011112222",
    };
    const subDomain = "subdomain";
    const prefix = "aesr-test";

    const globalCertificateStack = new GlobalCertificateStack(
      app,
      `${prefix}-global-certificate`,
      {
        env,
        crossRegionReferences: true,
        subDomain,
      },
    );

    const localCertificateStack = new LocalCertificateStack(
      app,
      `${prefix}-local-certificate`,
      {
        env,
        subDomain,
      },
    );

    const cognitoStack = new CognitoStack(app, `${prefix}-cognito`, {
      env,
      prefix,
      subDomain,
      certificate: globalCertificateStack.authCertificate,
      refreshTokenValidityDays: 30,
    });

    const apiGatewayStack = new ApiGatewayStack(app, `${prefix}-api-gateway`, {
      env,
      prefix,
      subDomain,
      certificate: localCertificateStack.apiCertificate,
      userPool: cognitoStack.userPool,
      configProviderLambdaName: "aesr-test-config-proivder",
    });

    const globalCertificateTemplate = Template.fromStack(
      globalCertificateStack,
    ).toJSON();
    expect(globalCertificateTemplate).toMatchSnapshot();

    const LocalCertificateTemplate = Template.fromStack(
      localCertificateStack,
    ).toJSON();
    expect(LocalCertificateTemplate).toMatchSnapshot();

    const cognitoTemplate = Template.fromStack(cognitoStack).toJSON();
    expect(cognitoTemplate).toMatchSnapshot();

    const apiGatewayTemplate = Template.fromStack(apiGatewayStack).toJSON();
    expect(apiGatewayTemplate).toMatchSnapshot();
  });
});
