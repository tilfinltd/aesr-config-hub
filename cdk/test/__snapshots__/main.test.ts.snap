// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All stacks default 1`] = `
{
  "Outputs": {
    "ExportsOutputRefAuthDomainCertificateABF6CF9E6A29C16A": {
      "Export": {
        "Name": "aesr-test-global-certificate:ExportsOutputRefAuthDomainCertificateABF6CF9E6A29C16A",
      },
      "Value": {
        "Ref": "AuthDomainCertificateABF6CF9E",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AuthDomainCertificateABF6CF9E": {
      "Properties": {
        "DomainName": "auth.subdomain.aesr.dev",
        "DomainValidationOptions": [
          {
            "DomainName": "auth.subdomain.aesr.dev",
            "HostedZoneId": "DUMMY",
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "aesr-test-global-certificate/AuthDomainCertificate",
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`All stacks default 2`] = `
{
  "Outputs": {
    "ExportsOutputRefApiDomainCertificate0C6AEA7E06B5AD39": {
      "Export": {
        "Name": "aesr-test-local-certificate:ExportsOutputRefApiDomainCertificate0C6AEA7E06B5AD39",
      },
      "Value": {
        "Ref": "ApiDomainCertificate0C6AEA7E",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ApiDomainCertificate0C6AEA7E": {
      "Properties": {
        "DomainName": "api.subdomain.aesr.dev",
        "DomainValidationOptions": [
          {
            "DomainName": "api.subdomain.aesr.dev",
            "HostedZoneId": "DUMMY",
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "aesr-test-local-certificate/ApiDomainCertificate",
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`All stacks default 3`] = `
{
  "Mappings": {
    "AWSCloudFrontPartitionHostedZoneIdMap": {
      "aws": {
        "zoneId": "Z2FDTNDATAQYW2",
      },
      "aws-cn": {
        "zoneId": "Z3RFFRIM2A3IF5",
      },
    },
  },
  "Outputs": {
    "ExportsOutputFnGetAttUserPool6BA7E5F2Arn686ACC00": {
      "Export": {
        "Name": "aesr-test-cognito:ExportsOutputFnGetAttUserPool6BA7E5F2Arn686ACC00",
      },
      "Value": {
        "Fn::GetAtt": [
          "UserPool6BA7E5F2",
          "Arn",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": {
      "DependsOn": [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-000011112222-ap-northeast-1",
          "S3Key": "17c16a3854838fd3ff4bda08146122a6701f33b9c86ae17f415ad0dc47a97544.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "AuthDomainRecord1650F0E5": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "UserPoolUserPoolDomainCloudFrontDomainName3917B242",
              "DomainDescription.CloudFrontDistribution",
            ],
          },
          "HostedZoneId": {
            "Fn::FindInMap": [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": "DUMMY",
        "Name": "auth.subdomain.aesr.dev.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "RootDomainRecordA96C29A1": {
      "Properties": {
        "HostedZoneId": "DUMMY",
        "Name": "subdomain.aesr.dev.",
        "ResourceRecords": [
          "127.0.0.1",
        ],
        "TTL": "1800",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "UserPool6BA7E5F2": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AccountRecoverySetting": {
          "RecoveryMechanisms": [
            {
              "Name": "verified_phone_number",
              "Priority": 1,
            },
            {
              "Name": "verified_email",
              "Priority": 2,
            },
          ],
        },
        "AdminCreateUserConfig": {
          "AllowAdminCreateUserOnly": true,
        },
        "EmailVerificationMessage": "The verification code to your new account is {####}",
        "EmailVerificationSubject": "Verify your new account",
        "Schema": [
          {
            "Mutable": true,
            "Name": "email",
            "Required": true,
          },
        ],
        "SmsVerificationMessage": "The verification code to your new account is {####}",
        "UserPoolName": "aesr-test-user-pool",
        "VerificationMessageTemplate": {
          "DefaultEmailOption": "CONFIRM_WITH_CODE",
          "EmailMessage": "The verification code to your new account is {####}",
          "EmailSubject": "Verify your new account",
          "SmsMessage": "The verification code to your new account is {####}",
        },
      },
      "Type": "AWS::Cognito::UserPool",
      "UpdateReplacePolicy": "Delete",
    },
    "UserPoolUserPoolClient40176907": {
      "Properties": {
        "AllowedOAuthFlows": [
          "code",
        ],
        "AllowedOAuthFlowsUserPoolClient": true,
        "AllowedOAuthScopes": [
          "openid",
          "email",
        ],
        "CallbackURLs": [
          "https://api.subdomain.aesr.dev/callback",
        ],
        "ClientName": "aesr-test-browser-extension",
        "GenerateSecret": false,
        "IdTokenValidity": 1080,
        "RefreshTokenValidity": 43200,
        "SupportedIdentityProviders": [
          "COGNITO",
        ],
        "TokenValidityUnits": {
          "IdToken": "minutes",
          "RefreshToken": "minutes",
        },
        "UserPoolId": {
          "Ref": "UserPool6BA7E5F2",
        },
      },
      "Type": "AWS::Cognito::UserPoolClient",
    },
    "UserPoolUserPoolDomain9F01E991": {
      "Properties": {
        "CustomDomainConfig": {
          "CertificateArn": {
            "Fn::ImportValue": "aesr-test-global-certificate:ExportsOutputRefAuthDomainCertificateABF6CF9E6A29C16A",
          },
        },
        "Domain": "auth.subdomain.aesr.dev",
        "UserPoolId": {
          "Ref": "UserPool6BA7E5F2",
        },
      },
      "Type": "AWS::Cognito::UserPoolDomain",
    },
    "UserPoolUserPoolDomainCloudFrontDomainName3917B242": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "UserPoolUserPoolDomainCloudFrontDomainNameCustomResourcePolicyF5EF8095",
      ],
      "Properties": {
        "Create": {
          "Fn::Join": [
            "",
            [
              "{"service":"CognitoIdentityServiceProvider","action":"describeUserPoolDomain","parameters":{"Domain":"",
              {
                "Ref": "UserPoolUserPoolDomain9F01E991",
              },
              ""},"physicalResourceId":{"id":"",
              {
                "Ref": "UserPoolUserPoolDomain9F01E991",
              },
              ""}}",
            ],
          ],
        },
        "InstallLatestAwsSdk": false,
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": {
          "Fn::Join": [
            "",
            [
              "{"service":"CognitoIdentityServiceProvider","action":"describeUserPoolDomain","parameters":{"Domain":"",
              {
                "Ref": "UserPoolUserPoolDomain9F01E991",
              },
              ""},"physicalResourceId":{"id":"",
              {
                "Ref": "UserPoolUserPoolDomain9F01E991",
              },
              ""}}",
            ],
          ],
        },
      },
      "Type": "Custom::UserPoolCloudFrontDomainName",
      "UpdateReplacePolicy": "Delete",
    },
    "UserPoolUserPoolDomainCloudFrontDomainNameCustomResourcePolicyF5EF8095": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "cognito-idp:DescribeUserPoolDomain",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "UserPoolUserPoolDomainCloudFrontDomainNameCustomResourcePolicyF5EF8095",
        "Roles": [
          {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`All stacks default 4`] = `
{
  "Outputs": {
    "aesrtestapiEndpoint0EA1DC97": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "aesrtestapi08E5B288",
            },
            ".execute-api.ap-northeast-1.",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "aesrtestapiDeploymentStageprod072FDD55",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ApiDomain2D27DFFC": {
      "Properties": {
        "DomainName": "api.subdomain.aesr.dev",
        "EndpointConfiguration": {
          "Types": [
            "REGIONAL",
          ],
        },
        "RegionalCertificateArn": {
          "Fn::ImportValue": "aesr-test-local-certificate:ExportsOutputRefApiDomainCertificate0C6AEA7E06B5AD39",
        },
      },
      "Type": "AWS::ApiGateway::DomainName",
    },
    "ApiDomainMapaesrtestapigatewayaesrtestapiDAC39E2906CBB9D3": {
      "Properties": {
        "DomainName": {
          "Ref": "ApiDomain2D27DFFC",
        },
        "RestApiId": {
          "Ref": "aesrtestapi08E5B288",
        },
        "Stage": {
          "Ref": "aesrtestapiDeploymentStageprod072FDD55",
        },
      },
      "Type": "AWS::ApiGateway::BasePathMapping",
    },
    "ApiDomainRecord355D35CF": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "ApiDomain2D27DFFC",
              "RegionalDomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::GetAtt": [
              "ApiDomain2D27DFFC",
              "RegionalHostedZoneId",
            ],
          },
        },
        "HostedZoneId": "DUMMY",
        "Name": "api.subdomain.aesr.dev.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "aesrtestapi08E5B288": {
      "Properties": {
        "Name": "aesr-test-api",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "aesrtestapiDeployment5AA56D05e325bb3e73176b9c7daacd131bf2f559": {
      "DependsOn": [
        "aesrtestapicallbackGETC98F9A4D",
        "aesrtestapicallbackB36D34ED",
        "aesrtestapiuserconfigGET432F3D56",
        "aesrtestapiuserconfigD48EE893",
        "aesrtestapiuser59805604",
        "cognitoAuthorizer9878DCCA",
      ],
      "Properties": {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": {
          "Ref": "aesrtestapi08E5B288",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "aesrtestapiDeploymentStageprod072FDD55": {
      "Properties": {
        "DeploymentId": {
          "Ref": "aesrtestapiDeployment5AA56D05e325bb3e73176b9c7daacd131bf2f559",
        },
        "RestApiId": {
          "Ref": "aesrtestapi08E5B288",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "aesrtestapicallbackB36D34ED": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "aesrtestapi08E5B288",
            "RootResourceId",
          ],
        },
        "PathPart": "callback",
        "RestApiId": {
          "Ref": "aesrtestapi08E5B288",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "aesrtestapicallbackGETC98F9A4D": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": {
          "IntegrationResponses": [
            {
              "ResponseTemplates": {
                "text/html": "<html>
            <body style="font-family:sans-serif;padding:36px">
             <p style="font-size:18px;line-height:2">
               Please keep this tab open and<br>
               <u>click on the <strong>AWS Extend Switch Roles</strong> icon</u>
               in your browser toolbar<br>
               to complete the connection to <strong>AESR Config Hub</strong>.
             </p>
            </body>
           </html>",
              },
              "StatusCode": "200",
            },
          ],
          "PassthroughBehavior": "WHEN_NO_MATCH",
          "RequestTemplates": {
            "application/json": "{"statusCode":200}",
          },
          "Type": "MOCK",
        },
        "MethodResponses": [
          {
            "ResponseModels": {
              "text/html": "Empty",
            },
            "StatusCode": "200",
          },
        ],
        "ResourceId": {
          "Ref": "aesrtestapicallbackB36D34ED",
        },
        "RestApiId": {
          "Ref": "aesrtestapi08E5B288",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "aesrtestapiuser59805604": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "aesrtestapi08E5B288",
            "RootResourceId",
          ],
        },
        "PathPart": "user",
        "RestApiId": {
          "Ref": "aesrtestapi08E5B288",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "aesrtestapiuserconfigD48EE893": {
      "Properties": {
        "ParentId": {
          "Ref": "aesrtestapiuser59805604",
        },
        "PathPart": "config",
        "RestApiId": {
          "Ref": "aesrtestapi08E5B288",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "aesrtestapiuserconfigGET432F3D56": {
      "Properties": {
        "AuthorizationType": "COGNITO_USER_POOLS",
        "AuthorizerId": {
          "Ref": "cognitoAuthorizer9878DCCA",
        },
        "HttpMethod": "GET",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": "arn:aws:apigateway:ap-northeast-1:lambda:path/2015-03-31/functions/arn:aws:lambda:ap-northeast-1:000011112222:function:aesr-test-config-proivder/invocations",
        },
        "ResourceId": {
          "Ref": "aesrtestapiuserconfigD48EE893",
        },
        "RestApiId": {
          "Ref": "aesrtestapi08E5B288",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "aesrtestapiuserconfigGETApiPermissionTestaesrtestapigatewayaesrtestapiDAC39E29GETuserconfig3373835B": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": "arn:aws:lambda:ap-northeast-1:000011112222:function:aesr-test-config-proivder",
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:aws:execute-api:ap-northeast-1:000011112222:",
              {
                "Ref": "aesrtestapi08E5B288",
              },
              "/test-invoke-stage/GET/user/config",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "aesrtestapiuserconfigGETApiPermissionaesrtestapigatewayaesrtestapiDAC39E29GETuserconfigE1F2EFE0": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": "arn:aws:lambda:ap-northeast-1:000011112222:function:aesr-test-config-proivder",
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:aws:execute-api:ap-northeast-1:000011112222:",
              {
                "Ref": "aesrtestapi08E5B288",
              },
              "/",
              {
                "Ref": "aesrtestapiDeploymentStageprod072FDD55",
              },
              "/GET/user/config",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "cognitoAuthorizer9878DCCA": {
      "Properties": {
        "IdentitySource": "method.request.header.Authorization",
        "Name": "CognitoAuthorizer",
        "ProviderARNs": [
          {
            "Fn::ImportValue": "aesr-test-cognito:ExportsOutputFnGetAttUserPool6BA7E5F2Arn686ACC00",
          },
        ],
        "RestApiId": {
          "Ref": "aesrtestapi08E5B288",
        },
        "Type": "COGNITO_USER_POOLS",
      },
      "Type": "AWS::ApiGateway::Authorizer",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
